ref/haskell/src/Codec/Binary/Bech32.hs

```diff
@@ -0,0 +1,128 @@
+module Codec.Binary.Bech32
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  , Word5(UnsafeWord5)
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor, testBit)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
+import Data.Word (Word8, Word32)
+import Data.Hashable (Hashable(..))
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+newtype Word5 = UnsafeWord5 Word8
+              deriving (Eq)
+
+instance Hashable Word5 where
+    hashWithSalt salt (UnsafeWord5 x) = hashWithSalt salt x
+
+word5 :: Word8 -> Word5
+word5 x = UnsafeWord5 (x .&. 31)
+
+fromWord5 :: Num a => Word5 -> a
+fromWord5 (UnsafeWord5 x) = fromIntegral x
+
```

Comment:
I looked at the `OddWord` library, and I'm not very keen on using it.  I immediately noticed that the `succ` and `pred` definitions are incorrect, and you've already noticed that it is missing `Ix` instances.

I can agree that all the instances I wrote out is maybe too much for removing a few uses of UnsafeWord5.

I suggest that you avoid `OddWord` and write

```
newType Word5 = UnsafeWord5 Word8 deriving (Eq, Ord)
```

Export the `word5` and `fromWord5` functions.  Add the `Ix` instance, and, at your discretion, maybe the `Num` instance, and leave it at that.  I suggest you stick the instances at the end of the file, because the instances are obvious, and no one will care to look at them.